### PR TITLE
Add TypeScript support (old; don't merge)

### PR DESCRIPTION
**Pull request by andrewsantarin**
I've got a working example of this using a test package using the TypeScript definitions in [`@andrewsantarin/staticify`](https://npmjs.com/packages/@andrewsantarin/staticify)

The working example: https://github.com/andrewsantarin/staticify-test

Clone it, install dependencies, see the source on a TypeScript-capable IDE and run locally to see it in action.

In all honesty, it's overkill to actually prove that the types work on an actual Node.js project, but I have other uses for it.

### CODE REVIEW

**Comment by andrewsantarin** [comment]
Okay, it looks like TypeScript on Windows absolutely hates it when I use wildcards in the `tsc` command.

image

I didn't see this earlier because used macOS, which didn't throw any errors because the terminal supports wildcards, I think? ([GitHub/typescript](https://github.com/Microsoft/TypeScript/issues/911), [Reddit](https://www.reddit.com/r/typescript/comments/7ozbcx/typescript_cli_does_it_support_globs_or_not/dsdmq4f?utm_source=share&utm_medium=web2x), [StackOverflow](https://stackoverflow.com/questions/12699781/how-can-i-pass-multiple-source-files-to-the-typescript-compiler)). Then I switched over to Windows and surely enough, found the problem.

I found out that there are two ways to solve this:

#### 1. Use tsconfig.json and let this file handle the wildcards for that, at the cost of add additional file.

On package.json
```json
{
  "scripts": {
    "tsc": "tsc"
  }
}
```

Then add a tsconfig.json file ([like so](https://github.com/Microsoft/dtslint#typestsconfigjson)) so that any TypeScript files under the /test folder will be picked up
```js
{
  "compilerOptions": {
    "module": "commonjs",
    "lib": [
      "es6"
    ],
    "strict": true,
    "noEmit": true,
    "baseUrl": ".",
    "paths": {
      "staticify": [ // Here be it
        "."
      ]
    }
  }
}
```

#### 2. Explicitly mention which files to pass to `tsc`, at the cost of manually adding a new file each time we need it.

On package.json
```
{
  "scripts": {
    "tsc": "tsc --esModuleInterop --noEmit test/index.ts"
  }
}
```

**Comment by rakeshpai** [comment]
Thanks for this. I'll review over the weekend.

**Comment by andrewsantarin** [comment]
All right, @rakeshpai.

Do let me know which of the two approaches you'd rather go with. I took the 2nd one; the typecheck is simple and doesn't really need tsconfig.json yet.

**Comment by rakeshpai** [comment]
Hi @andrewsantarin I like the 2nd approach too. Should be good enough for the time being. 

Just a thought - rather than creating a .d.ts file, wouldn't it be possible to use JSDoc comments and use tsc to generate the declaration files? Example: https://dev.to/open-wc/generating-typescript-definition-files-from-javascript-5bp2 (I've never used this technique myself.)

The benefit of this is that it's going to be easier for a JS developer to maintain, assuming no familiarity with TS.

**Comment by andrewsantarin** [comment]
[The article you mentioned](https://dev.to/open-wc/generating-typescript-definition-files-from-javascript-5bp2) concludes that TypeScript has not yet made this an official feature yet. 

While, yeah, we could use JSDoc to generate to generate `*.d.ts`, we require a tool to do that for us: 
- [a fork of `typescript`](https://www.npmjs.com/package/typescript-temporary-fork-for-jsdoc) exists, but it has **no guarantee** that it'll be supported in the future (with the fellow engineers I work with, this is a red flag).
- [`phaser`](https://npmjs.com/packages/phaser) uses [its owntsgen tool](https://github.com/photonstorm/phaser/tree/master/scripts/tsgen), which looks like a lot of work and only pays off if the source code is huge. We only have here, like, what, one `index.js`?

Also, consider the fact that you have a dependency with modules such as `send`. I'm not sure how that works with JSDoc comments. Hopefully, you know how to typify `options.sendOptions` as `SendOptions` correctly on JSDoc, because I sure as hell don't. 😅

**tl;dr:** It's possible, but a resounding "impractical given the hands we have", in my opinion.

**Comment by andrewsantarin** [comment]
@rakeshpai 

[The article you mentioned](https://dev.to/open-wc/generating-typescript-definition-files-from-javascript-5bp2) concludes that TypeScript has not yet made this an official feature yet. 

While, yeah, we could use JSDoc to generate to generate `*.d.ts`, we require a tool to do that for us: 
- [a fork of `typescript`](https://www.npmjs.com/package/typescript-temporary-fork-for-jsdoc) exists, but it has **no guarantee** that it'll be supported in the future (with the fellow engineers I work with, this is a red flag).
- [`phaser`](https://npmjs.com/packages/phaser) uses [its owntsgen tool](https://github.com/photonstorm/phaser/tree/master/scripts/tsgen), which looks like a lot of work and only pays off if the source code is huge. We only have here, like, what, one `index.js`?

Also, consider the fact that you have a dependency with modules such as `send`. I'm not sure how that works with JSDoc comments. Hopefully, you know how to typify `options.sendOptions` as `SendOptions` correctly on JSDoc, because I sure as hell don't. 😅

**tl;dr:** It's possible, but a resounding "impractical given the hands we have", in my opinion.

**Comment by XhmikosR** [comment]
@andrewsantarin you need to double quote the globs in npm scripts. Never forget this :P

**Comment by andrewsantarin** [comment]
> @andrewsantarin you need to double quote the globs in npm scripts. Never forget this :P

Sorry, @rakeshpai, I don't understand what you mean. Where?

**Comment by XhmikosR** [comment]
https://github.com/errorception/staticify/pull/49#issuecomment-559632921

The reason is the missing double quotes.

**Comment by rakeshpai** [comment]
https://github.com/errorception/staticify/pull/49#issuecomment-560959429

Ok, seems fair. This PR LGTM. @XhmikosR does this look fine to you?

**Comment by XhmikosR** [comment]
I don't use Typescript myself so I can't comment on the change itself.

**Comment by XhmikosR** [comment]
Also, I'm not sure about the added test. Why do we need it? I'd rather have less scripts personally.

**Comment by andrewsantarin** [comment]
@XhmikosR It's there so that if the machine that's running the tests doesn't have, let's say a TypeScript-compatible editor like VSCode, you can still verify the types work. Personally, I don't need it, so if simpler scripts is seriously the way to go, then ping me so that I can remove the `tsc` script.

Also, @rakeshpai @XhmikosR, please excuse my being late! Hope you're still keen on accepting this PR.
